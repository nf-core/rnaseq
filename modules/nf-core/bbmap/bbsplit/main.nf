process BBMAP_BBSPLIT {
    tag "$meta.id"
    label 'process_high'
    label 'error_retry'

    conda "${moduleDir}/environment.yml"
    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://depot.galaxyproject.org/singularity/bbmap:39.01--h5c4e2a8_0':
        'biocontainers/bbmap:39.01--h5c4e2a8_0' }"

    input:
    tuple val(meta), path(reads)
    path  index
    path  primary_ref
    tuple val(other_ref_names), path(other_ref_paths)
    val   only_build_index

    output:
    path "bbsplit"                            , optional:true, emit: index
    tuple val(meta), path('*primary*fastq.gz'), optional:true, emit: primary_fastq
    tuple val(meta), path('*fastq.gz')        , optional:true, emit: all_fastq
    tuple val(meta), path('*txt')             , optional:true, emit: stats
    path "versions.yml"                       , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"

    def avail_mem = 3072
    if (!task.memory) {
        log.info '[BBSplit] Available memory not known - defaulting to 3GB. Specify process memory requirements to change this.'
    } else {
        avail_mem = (task.memory.mega*0.8).intValue()
    }

    def other_refs = []
    other_ref_names.eachWithIndex { name, index ->
        other_refs << "ref_${name}=${other_ref_paths[index]}"
    }
    if (only_build_index) {
        if (primary_ref && other_ref_names && other_ref_paths) {
            """
            bbsplit.sh \\
                -Xmx${avail_mem}M \\
                ref_primary=$primary_ref \\
                ${other_refs.join(' ')} \\
                path=bbsplit \\
                threads=$task.cpus \\
                $args

            cat <<-END_VERSIONS > versions.yml
            "${task.process}":
                bbmap: \$(bbversion.sh | grep -v "Duplicate cpuset")
            END_VERSIONS
            """
        } else {
            log.error 'ERROR: Please specify as input a primary fasta file along with names and paths to non-primary fasta files.'
        }
    } else {
        def index_files = ''
        if (index) {
            index_files = "path=$index"
        } else if (primary_ref && other_ref_names && other_ref_paths) {
            index_files = "ref_primary=${primary_ref} ${other_refs.join(' ')}"
        } else {
            log.error 'ERROR: Please either specify a BBSplit index as input or a primary fasta file along with names and paths to non-primary fasta files.'
        }
        def fastq_in  = meta.single_end ? "in=${reads}" : "in=${reads[0]} in2=${reads[1]}"
        def fastq_out = meta.single_end ? "basename=${prefix}_%.fastq.gz" : "basename=${prefix}_%_#.fastq.gz"
        """
        bbsplit.sh \\
            -Xmx${avail_mem}M \\
            $index_files \\
            threads=$task.cpus \\
            $fastq_in \\
            $fastq_out \\
            refstats=${prefix}.stats.txt \\
            $args

        cat <<-END_VERSIONS > versions.yml
        "${task.process}":
            bbmap: \$(bbversion.sh | grep -v "Duplicate cpuset")
        END_VERSIONS
        """
    }

}
