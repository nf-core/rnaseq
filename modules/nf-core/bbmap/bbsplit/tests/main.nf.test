nextflow_process {

    name "Test Process BBMAP_BBSPLIT"
    script "../main.nf"
    process "BBMAP_BBSPLIT"
    tag "modules"
    tag "modules_nfcore"
    tag "bbmap"
    tag "bbmap/bbsplit"

    test("sarscov2_se_fastq_fasta_chr22_fasta") {

        setup {

            run("BBMAP_BBSPLIT", alias: "BBMAP_BBSPLIT_INDEX") {
                script "../main.nf"
                process {
                    """
                    input[0] = [[:],[]]
                    input[1] = []
                    input[2] = Channel.of(file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/genome.fasta', checkIfExists: true))
                    input[3] = Channel.of([
                        [ 'human' ], // meta map
                        file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/chr22/sequence/chr22_23800000-23980000.fa', checkIfExists: true)
                    ])
                    input[4] = true
                    """
                }
            }
        }

        when {
            params {
                outdir   = "$outputDir"
            }
            process {
                    """
                    input[0] = Channel.of([
                        [ id:'test', single_end:true ], // meta map
                        file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/fastq/test_1.fastq.gz', checkIfExists: true)
                    ])
                    input[1] = BBMAP_BBSPLIT_INDEX.out.index
                    input[2] = Channel.of(file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/genome.fasta', checkIfExists: true))
                    input[3] = Channel.of([
                        [ 'human' ], // meta map
                        file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/chr22/sequence/chr22_23800000-23980000.fa', checkIfExists: true)
                    ])
                    input[4] = true
                    """
                }
        }

        then {
            def filesToExclude = [
                "info.txt",
                "reflist.txt",
                "scaffolds.txt.gz",
                "summary.txt"
            ]

            def outputFiles = []
            def outputDirectory = new File(process.out.index[0])
            outputDirectory.eachFileRecurse { file ->
                if (file.isFile()){
                    outputFiles << file
                    }
            }

            def filesExist = filesToExclude.every { excludeName ->
                outputFiles.any { file ->
                    file.getName().endsWith(excludeName) && file.exists()
                }
            }

            def filteredFiles = outputFiles
                .findAll { file ->
                    !filesToExclude.any { excludeName ->
                        file.getName().endsWith(excludeName)
                    }
                }.sort{ it.getName()}

            assertAll(
                { assert process.success },
                { assert snapshot(filteredFiles).match("bbsplit_index_filtered_files")},
                { assert filesExist : "One or more files to exclude do not exist" },
                { assert snapshot(process.out.versions).match("versions")}
            )
        }
    }
}
