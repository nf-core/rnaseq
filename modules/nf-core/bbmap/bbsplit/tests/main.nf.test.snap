{
    "bbsplit_index_filtered_files": {
        "content": [
            [
                "chr1.chrom.gz:md5,8fec4c63ec642613ad10adf4cc2e6ade",
                "chr1_index_k13_c13_b1.block:md5,385913c1e84b77dc7bf36288ee1c8706",
                "chr1_index_k13_c13_b1.block2.gz:md5,2556b45206835a0ff7078d683b5fd6e2",
                "merged_ref_9222711925172838098.fa.gz:md5,983cef447fb28394b88a5b49b3579f0c",
                "namelist.txt:md5,45e7a4cdc7a11a39ada56844ca3a1e30"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-02-01T17:15:59.705013452"
    },
    "versions": {
        "content": [
            [
                "versions.yml:md5,cb7f0e697ab2537f8ced951bfade90d4"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-02-01T17:11:22.441753642"
    }
}
